src/lib.rs

```diff
@@ -390,140 +380,128 @@ impl<'a, T: Send + Sync> Iterator for Iter<'a, T> {
                 }
             }
 
-            if self.bucket != 0 {
-                self.bucket_size <<= 1;
-            }
-            self.bucket += 1;
-
-            self.index = 0;
+            self.next_bucket();
         }
         None
     }
-
-    fn size_hint(&self) -> (usize, Option<usize>) {
-        let total = self.thread_local.values.load(Ordering::Acquire);
-        (total - self.yielded, None)
-    }
-}
-impl<T: Send + Sync> FusedIterator for Iter<'_, T> {}
-
-struct RawIterMut<T: Send> {
-    remaining: usize,
-    buckets: [*mut Entry<T>; BUCKETS],
-    bucket: usize,
-    bucket_size: usize,
-    index: usize,
-}
-
-impl<T: Send> Iterator for RawIterMut<T> {
-    type Item = *mut MaybeUninit<T>;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        if self.remaining == 0 {
+    fn next_mut<'a, T: Send>(
+        &mut self,
+        thread_local: &'a mut ThreadLocal<T>,
+    ) -> Option<&'a mut Entry<T>> {
+        if *thread_local.values.get_mut() == self.yielded {
             return None;
         }
 
         loop {
-            let bucket = unsafe { *self.buckets.get_unchecked(self.bucket) };
+            let bucket = unsafe { thread_local.buckets.get_unchecked_mut(self.bucket) };
+            let bucket = *bucket.get_mut();
 
             if !bucket.is_null() {
                 while self.index < self.bucket_size {
                     let entry = unsafe { &mut *bucket.add(self.index) };
                     self.index += 1;
                     if *entry.present.get_mut() {
-                        self.remaining -= 1;
-                        return Some(entry.value.get());
+                        self.yielded += 1;
+                        return Some(entry);
                     }
                 }
             }
 
-            if self.bucket != 0 {
-                self.bucket_size <<= 1;
-            }
-            self.bucket += 1;
+            self.next_bucket();
+        }
+    }
 
-            self.index = 0;
+    fn next_bucket(&mut self) {
+        if self.bucket != 0 {
+            self.bucket_size <<= 1;
         }
+        self.bucket += 1;
+        self.index = 0;
     }
 
-    fn size_hint(&self) -> (usize, Option<usize>) {
-        (self.remaining, Some(self.remaining))
+    fn size_hint<T: Send>(&self, thread_local: &ThreadLocal<T>) -> (usize, Option<usize>) {
+        let total = thread_local.values.load(Ordering::Acquire);
```

Comment:
This can be `Relaxed`.

Reply:
I'm not an export on atomic orderings at all, but wouldn't this allow the processor to reorder getting the next value before loading this one, rendering the `size_hint` too high? Or do the atomic loads in `next` prevent that?

Reply:
Actually you are correct.